context.go

```diff
@@ -0,0 +1,133 @@
+	// Cumulated overall run time - in nanoseconds - for this context.
+	totalOverallRuntimeNs atomic.Uint64
+	// Cumulated timeout count for this context.
+	timeoutCount atomic.Uint64
+}
+
+// Run encodes the given addressesToData values and runs them against the WAF rules within the given
+// timeout value. It returns the matches as a JSON string (usually opaquely used) along with the corresponding
+// actions in any. In case of an error, matches and actions can still be returned, for instance in the case of a
+// timeout error. Errors can be tested against the RunError type.
+func (context *Context) Run(addressesToData map[string]any, timeout time.Duration) (matches []byte, actions []string, err error) {
+	if len(addressesToData) == 0 {
+		return
+	}
+
+	now := time.Now()
+	defer func() {
+		dt := time.Since(now)
+		context.totalOverallRuntimeNs.Add(uint64(dt.Nanoseconds()))
+	}()
+
+	encoder := encoder{
+		stringMaxSize:    wafMaxStringLength,
+		containerMaxSize: wafMaxContainerSize,
+		objectMaxDepth:   wafMaxContainerDepth,
+	}
+	obj, err := encoder.Encode(addressesToData)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// ddwaf_run cannot run concurrently and the next append write on the context state so we need a mutex
+	context.mutex.Lock()
+	defer context.mutex.Unlock()
+	defer context.cgoRefs.append(encoder.cgoRefs)
+
+	return context.run(obj, timeout, &encoder.cgoRefs)
+}
+
+func (context *Context) run(obj *wafObject, timeout time.Duration, cgoRefs *cgoRefPool) ([]byte, []string, error) {
+	// RLock the handle to safely get read access to the WAF handle and prevent concurrent changes of it
+	// such as a rules-data update.
+	context.handle.mutex.RLock()
+	defer context.handle.mutex.RUnlock()
+
+	result := new(wafResult)
+	defer wafLib.wafResultFree(result)
+
+	ret := wafLib.wafRun(context.cContext, obj, result, uint64(timeout/time.Microsecond))
+
+	context.totalRuntimeNs.Add(result.total_runtime)
+	matches, actions, err := unwrapWafResult(ret, result)
+	if err == ErrTimeout {
+		context.timeoutCount.Inc()
+	}
+
+	return matches, actions, err
+}
+
+func unwrapWafResult(ret wafReturnCode, result *wafResult) (matches []byte, actions []string, err error) {
+	if result.timeout > 0 {
+		err = ErrTimeout
+	}
+
+	if ret == wafOK {
+		return nil, nil, err
+	}
+
+	if ret != wafMatch {
+		return nil, nil, goRunError(ret)
+	}
+
+	if result.data != 0 {
+		matches = []byte(gostring(cast[byte](result.data)))
+	}
+
+	if size := result.actions.size; size > 0 {
+		actions = decodeActions(result.actions.array, uint64(size))
+	}
+
+	return matches, actions, err
+}
+
+// Close calls handle.closeContext which calls ddwaf_context_destroy and maybe also close the handle if it in termination state.
+func (context *Context) Close() {
+	// Needed to make sure the garbage collector does not throw the values send to the WAF earlier than necessary
+	keepAlive(context.cgoRefs.stringRefs)
+	keepAlive(context.cgoRefs.arrayRefs)
+	context.handle.closeContext(context)
```

Comment:
1. I don't see any value for `closeContext` which spreads to context closure into an handle method
2. The `keepAlive` calls don't make sense to me here, because nothing special happens before.

So here is my suggestion:
```suggestion
	defer context.handle.closeContext(context)
	// Needed to make sure the garbage collector does not throw the values send to the WAF earlier than necessary
	wafLib.wafContextDestroy(context.cContext)
	// Keep the Go pointer references until the end of the context
	keepAlive(context.cgoRefs)
	// The context is no longer used so we can try releasing the Go pointer references asap by nulling them
	context.cgoRefs = cgoRefPool{}
```

If applied, renaming `closeContext` sounds better too.

Reply:
I was against the idea because of owernship. Just like why NewContext() is within the handle scope, I think that destroying the context should also be done via the handle. This idea already became a little flaky after you told me to change NewContext() from a method to a function but yeah

Reply:
Part of this proposal still apply even with closeContext:
```suggestion
	defer context.handle.closeContext(context)
	// Keep the Go pointer references until the end of the context
	keepAlive(context.cgoRefs)
	// The context is no longer used so we can try releasing the Go pointer references asap by nulling them
	context.cgoRefs = cgoRefPool{}
```